build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1185,6 +1417,156 @@ message ExecuteRequest {
+  // indicates a directory that the `Action` was not predicted to use
+  // from the client filesystem. The client is requested to enumerate
+  // directory metadata and respond in
+  // [StreamedDynamicExecutionData.dynamic_directory_metadata_response][build.bazel.remote.execution.StreamedDynamicExecutionData].
+  // The client MAY include individual file digests but this is not required,
+  // the worker will request individual file hashes as needed using
+  // `dynamic_file_hash_path`.
+  // Depending on the server and client implementation this may be a
+  // relative or fully qualified path.
+  string dynamic_directory_content_path = 4;
+
+  // A file path for which the worker does not have a hash. This indicates
+  // a file that the `Action` was not predicted to use from the client
+  // filesystem. The client is requested to hash the file and respond in
+  // [StreamedDynamicExecutionData.dynamic_file_hash_response][build.bazel.remote.execution.StreamedDynamicExecutionData].
+  // Depending on the server and client implementation this may be a
+  // relative or fully qualified path.
+  string dynamic_file_hash_path = 5;
+
+  // A file path and expected content hash that the worker does not have in
+  // the CAS. This indicates file content that the `Action` was not predicted
+  // to use from the client, or content that was subsequently removed from
+  // the CAS. The client is requested to upload the file content into the
+  // CAS and respond in 
+  // [StreamedDynamicExecutionData.dynamic_file_content_response][build.bazel.remote.execution.StreamedDynamicExecutionData].
+  DynamicFileContentRequest dynamic_file_content_request = 6;
+}
+
+// A streaming request message for
+// [Execution.ExecuteWithDynamicRetrieval][build.bazel.remote.execution.v2.Execution.ExecuteWithDynamicRetrieval].
+message StreamedDynamicExecutionData {
+  // Sent in response to a dynamic directory metadata request.
+  message DynamicDirectoryMetadataResponse {
+    // The directory path, copied from the directory metadata request.
+    string path = 1;
+
+    // Metadata for the directory. [DirectoryNode][build.bazel.remote.execution.v2.Execution.DirectoryNode]
+    // instances in this field MAY omit digest reference information, as
+    // the directory name is the required information needed to feed into
+    // the filesystem, and the contents of the subdirectory contained in the
+    // `DirectoryNode` are not required to be in the CAS. If the
+    // `DirectoryNode` digest is sent to the worker, the referenced `Directory`
+    // information MUST have been previously uploaded to the CAS.
+    Directory directory_metadata = 2;
+  }
+
+  // Sent in response to a dynamic file hash request.
+  message DynamicFileHashResponse {
+    // The file path, copied from the file hash request.
+    string path = 1;
+    
+    // Set to true if the requested file path does not exist.
+    bool file_does_not_exist = 2;
```

Comment:
How can this happen? The worker shouldn't be able to request information about a file unless the `FileNode` is present, right?

Reply:
In actual practice the client's filesystem is sometimes changed underneath execution, particularly in a highly parallel build using build engines (like MSBuild) whose build DAG is not strongly specified.